File: CSES/ShortestRoutesII.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const long long MAX = 1e15;

void floydWarshall(int n, vector<vector<long long>>& dist) {
    for (int k = 0; k < n; ++k) {
        for (int i = 0; i < n; ++i) {
            if (dist[i][k] < MAX) {
                for (int j = 0; j < n; ++j) {
                    if (dist[k][j] < MAX) {
                        dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);
                    }
                }
            }
        }
    }
}

int main() {
    int n, m, q;
    cin >> n >> m >> q;

    vector<vector<long long>> dist(n, vector<long long>(n, MAX));

    for (int i = 0; i < n; ++i)
        dist[i][i] = 0;

    for (int i = 0; i < m; ++i) {
        int u, v;
        long long w;
        cin >> u >> v >> w;
        --u; --v;
        dist[u][v] = min(dist[u][v], w);
        dist[v][u] = min(dist[v][u], w);
    }

    floydWarshall(n, dist);

    for (int i = 0; i < q; ++i) {
        int u, v;
        cin >> u >> v;
        --u; --v;
        if (dist[u][v] < MAX) {
            cout << dist[u][v] << '\n';
        } else {
            cout << -1 << '\n';
        }
    }

    return 0;
}

File: CSES/Python/ShortestRoutesII.py
import sys
from heapq import *
MAX = 10**15
 
def floydWarshall(n, dist):
    for k in range(n):
        for i in range(n):
            if dist[i][k] < MAX:
                for j in range(n):
                    if dist[j][k] < MAX:
                        dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])
    
def solve():
    data = list(map(int, sys.stdin.read().split()))
    n, m, q, idx = data[0], data[1], data[2], 3
    dist = [[MAX for _ in range(n)] for _ in range(n)]
    for _ in range(m):
        u, v, w = data[idx] - 1, data[idx + 1] - 1, data[idx + 2]
        dist[u][v] = dist[v][u] = w
        idx += 3
    
    floydWarshall(n, dist)
    for _ in range(q):
        u, v = data[idx] - 1, data[idx + 1] - 1
        print(dist[u][v] if dist[u][v] != MAX else -1)
        idx += 2
 
if __name__ == "__main__":
    solve()